### PR TITLE
snapshots: Refuse snapshots of running VMs with VFIO host devices

### PR DESCRIPTION
**Pull request by mvollmer**

### CODE REVIEW

**Comment by mvollmer** [comment]
I do not know how to test this. We would need a running machine with a VFIO device, which means we need to fake one convincingly enough that qemu will actually run with it. Or find a real one.

**Comment by mvollmer** [comment]
I think it's fine not to have automated tests for this.

File: src/components/vm/snapshots/vmSnapshotsCard.jsx [hunk 1]
```diff
@@ -53,11 +53,20 @@ export const VmSnapshotsActions = ({ vm, config, storagePools }) => {
                                           vm={vm} />);
     }
 
-    return (
-        <Button id={`${id}-add-snapshot-button`} variant="secondary" onClick={open}>
+    let excuse = null;
+    if (vm.state != 'shut off' && vmHasVFIOHostDevs(vm))
+        excuse = _("Creating snapshots of VMs with VFIO devices is not supported while they are running.");
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCard.jsx:58]: This added line is not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCard.jsx [hunk 2]
```diff
@@ -53,11 +53,20 @@ export const VmSnapshotsActions = ({ vm, config, storagePools }) => {
...
+    const button = (
+        <Button id={`${id}-add-snapshot-button`}
+                variant="secondary"
+                onClick={open}
+                isAriaDisabled={!!excuse}>
             {_("Create snapshot")}
         </Button>
     );
+
+    return excuse ? <Tooltip content={excuse}>{button}</Tooltip> : button;
```
Comment (cockpituous) [inline src/components/vm/snapshots/vmSnapshotsCard.jsx:69]: This added line is not executed by any test.

File: src/libvirt-xml-parse.js [hunk 3]
```diff
@@ -640,8 +640,11 @@ export function parseDumpxmlForHostDevices(devicesElem) {
         for (let i = 0; i < hostdevElems.length; i++) {
             const hostdevElem = hostdevElems[i];
             const bootElem = getSingleOptionalElem(hostdevElem, 'boot');
+            const driverElem = getSingleOptionalElem(hostdevElem, 'driver');
             const type = hostdevElem.getAttribute('type');
             const mode = hostdevElem.getAttribute('mode');
+            const driver = driverElem?.getAttribute('name');
```
Comment (cockpituous) [inline src/libvirt-xml-parse.js:646]: This added line is not executed by any test.

File: src/libvirt-xml-parse.js [hunk 4]
```diff
@@ -720,6 +725,7 @@ export function parseDumpxmlForHostDevices(devicesElem) {
                     type,
                     mode,
                     bootOrder: bootElem?.getAttribute('order'),
+                    driver,
```
Comment (cockpituous) [inline src/libvirt-xml-parse.js:728]: This added line is not executed by any test.

File: src/libvirt-xml-parse.js [hunk 5]
```diff
@@ -759,6 +766,7 @@ export function parseDumpxmlForHostDevices(devicesElem) {
...
```
Comment (cockpituous) [inline src/libvirt-xml-parse.js:769]: This added line is not executed by any test.

File: src/libvirt-xml-parse.js [hunk 6]
```diff
@@ -776,6 +784,7 @@ export function parseDumpxmlForHostDevices(devicesElem) {
...
```
Comment (cockpituous) [inline src/libvirt-xml-parse.js:787]: This added line is not executed by any test.

File: src/libvirt-xml-parse.js [hunk 7]
```diff
@@ -806,6 +815,7 @@ export function parseDumpxmlForHostDevices(devicesElem) {
...
```
Comment (cockpituous) [inline src/libvirt-xml-parse.js:818]: This added line is not executed by any test.

File: src/components/vm/snapshots/vmSnapshotsCard.jsx [hunk 8]
```diff
@@ -53,11 +53,20 @@ export const VmSnapshotsActions = ({ vm, config, storagePools }) => {
...
-        <Button id={`${id}-add-snapshot-button`} variant="secondary" onClick={open}>
+    let excuse = null;
+    if (vm.state != 'shut off' && vmHasVFIOHostDevs(vm))
+        excuse = _("Creating snapshots of VMs with VFIO devices is not supported while they are running.");
+
+    const button = (
+        <Button id={`${id}-add-snapshot-button`}
+                variant="secondary"
+                onClick={open}
+                isAriaDisabled={!!excuse}>
```
[the remark this answers is not in the document]
Comment (mvollmer) [reply src/components/vm/snapshots/vmSnapshotsCard.jsx:64]: I don't know... I copy/pasted this from bootOrder.jsx, and with `isDisabled`, the tooltip doesn't trigger.

**Review by jelly — APPROVED** [review]
I imagine we can't test VFIO, so LGTM

**Pull request merged**